contributors/design-proposals/raw-block-pv.md

```diff
@@ -0,0 +1,572 @@
+
+BACKGROUND: The admin wishes to create a storage class that will identify pre-provisioned block PVs based on a user's PVC request for volumeMode: Block. 
+
+WORKFLOW: 
+
+ADMIN:
+
+```
+kind: StorageClass
+apiVersion: storage.k8s.io/v1
+metadata:
+  name: block-volume
+provisioner: no-provisioning 
+parameters:
+```
+* Sample of pre-created volume definition:
+
+```
+apiVersion: v1
+kind: PersistentVolume
+metadata:
+ name: pv-block-volume
+spec:
+ volumeMode: Block
+ storageClassName: block-volume
+ capacity:
+   storage: 35Gi
+ accessModes:
+   - ReadWriteOnce
+ local:
+    path: /dev/xvdc
+```
+## [FUTURE] UC5: 
+
+DESCRIPTION: StorageClass with dynamically created volumes 
+
+BACKGROUND: The admin wishes to create a storage class that will dynamically create block PVs based on a user's PVC request for volumeMode: Block. The admin desires the volumes be created dynamically and deleted when the PV definition is deleted. 
+
+WORKFLOW:
+
+ADMIN:
+
+```
+kind: StorageClass
+apiVersion: storage.k8s.io/v1
+metadata:
+  name: local-fast
+provisioner: kubernetes.io/local-block-ssd
+parameters:
+  volumeMode: Block #suggested value - this is plugin/provisioner dependent
+```
+
+***This has implementation details that have yet to be determined. It is included in this proposal for completeness of design ****
+
+## UC6: 
+
+DESCRIPTION: The developer wishes to request a block device via a Storage Class.
+
+WORKFLOW:
+
+USER: 
+
+```
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+ name: pvc-local-block
+spec:
+ volumeMode: Block
+ storageClassName: local-fast
+ accessModes:
+  - ReadWriteOnce
+ resources:
+   requests:
+     storage: 10Gi
+```  
+
+## UC7: 
+
+DESCRIPTION: Admin creates network raw block devices
+
+BACKGROUND: Admin wishes to pre-create Persistent Volumes in GCE as raw block devices
+
+WORKFLOW:
+
+ADMIN:
+
+```
+apiVersion: "v1"
+kind: "PersistentVolume"
+metadata:
+  name: gce-disk-1
+  annotations:
+    volume.beta.kubernetes.io/mount-options: "discard"
+Spec:
+  volumeMode: Block
+  capacity:
+    storage: "10Gi"
+  accessModes:
+    - "ReadWriteOnce"
+  gcePersistentDisk:
+    fsType: "block" #this is a suggestion, it is plugin 
+    pdName: "gce-disk-1"
+```
+
+***If admin specifies volumeMode: Block + fstype: ext4 then they would have the default behavior of files on block ***
+***fsType values will be provisioner dependent. Block is suggested for development simplicity. Since the PVC object is passed
+   to the provisioner, it will be responsible for validating and handling whether or not it supports the volumeMode being passed ***
+
+## UC8: 
+
+DESCRIPTION: 
+
+A developer wishes to enable their application to use a raw block device as an inline volume in the pod. 
+
+WORKFLOW:
+
+USER:
+
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  name: my-db
+spec:
+    containers:
+    - name: mysql
+      image: mysql
+      volumeDevices:
+      - name: my-db-data
+        devicePath: /var/lib/mysql/data
```

Comment:
/dev/sdb

Reply:
no, this is correct

Reply:
@erinboyd 
Could you add comments or notes about the "data" file?
The "data" is actually device file and user can access raw block device through this device file.
This is confusing, so additional explanation would be better here.

Reply:
I think the others are trying to say that typical convention is to put all block devices under "/dev", instead of having it under "/var/lib", which is typically associated with filesystems.  So having a devicePath of "/dev/mysql-data" may be less misleading.